From 810a6fecfaa329ef727d00901170e4abf589e523 Mon Sep 17 00:00:00 2001
From: Peter W Morreale <pmorreal@brocade.com>
Date: Fri, 21 Feb 2014 10:52:52 -0700
Subject: [PATCH 1/5] Add a local ip range option.

This patch adds the range field to the LNS struct.  It is initialized
to NULL.  It also adds a handler to parse the option.

Signed-off-by: Peter W Morreale <pmorreal@brocade.com>
---
 file.c | 24 ++++++++++++++++++++++++
 file.h |  1 +
 2 files changed, 25 insertions(+)

diff --git a/file.c b/file.c
index c93dba8..9f28d74 100644
--- a/file.c
+++ b/file.c
@@ -103,6 +103,7 @@ struct lns *new_lns ()
     tmp->hostname[0] = 0;
     tmp->entname[0] = 0;
     tmp->range = NULL;
+    tmp->localrange = NULL;
     tmp->assign_ip = 1;                /* default to 'yes' */
     tmp->lacs = NULL;
     tmp->passwdauth = 0;
@@ -967,6 +968,28 @@ int set_iprange (char *word, char *value, int context, void *item)
     return 0;
 }
 
+int set_localiprange (char *word, char *value, int context, void *item)
+{
+    struct lns *lns = (struct lns *) item;
+    switch (context & ~CONTEXT_DEFAULT)
+    {
+    case CONTEXT_LNS:
+        break;
+    default:
+        snprintf (filerr, sizeof (filerr), "'%s' not valid in this context\n",
+                  word);
+        return -1;
+    }
+    lns->localrange = set_range (word, value, lns->localrange);
+    if (!lns->localrange)
+        return -1;
+#ifdef DEBUG_FILE
+    l2tp_log (LOG_DEBUG, "range start = %x, end = %x, sense=%ud\n",
+         ntohl (lns->range->start), ntohl (lns->range->end), lns->range->sense);
+#endif
+    return 0;
+}
+
 int set_lac (char *word, char *value, int context, void *item)
 {
     struct lns *lns = (struct lns *) item;
@@ -1511,6 +1534,7 @@ struct keyword words[] = {
     {"no lac", &set_lac},
     {"assign ip", &set_assignip},
     {"local ip", &set_localaddr},
+    {"local ip range", &set_localiprange},
     {"remote ip", &set_remoteaddr},
     {"defaultroute", &set_defaultroute},
     {"length bit", &set_lbit},
diff --git a/file.h b/file.h
index 174cd53..9f7b73d 100644
--- a/file.h
+++ b/file.h
@@ -81,6 +81,7 @@ struct lns
     char entname[STRLEN];       /* Name of this entry */
     struct iprange *lacs;       /* Hosts permitted to connect */
     struct iprange *range;      /* Range of IP's we provide */
+    struct iprange *localrange; /* Range of local IP's we provide */
     int assign_ip;              /* Do we actually provide IP addresses? */
     int passwdauth;             /* Authenticate by passwd file? (or PAM) */
     int pap_require;            /* Require PAP auth for PPP */
-- 
1.8.4.5

